docs/performance-guide.md

```diff
@@ -25,150 +54,205 @@ guide intends to answer the following questions:
+The transport type, [Websockets](https://en.wikipedia.org/wiki/WebSocket), [Sever-Sent-Event](https://en.wikipedia.org/wiki/Server-sent_events), or [Long-Polling](https://en.wikipedia.org/wiki/Push_technology) is another factor which impacts performance. Websockets is a bidirectional and full-duplex communication protocol over a single TCP connection. However, Sever-Sent-Event is unidirectional protocol to push message from server to client. Long-Polling requires the clients to poll information from server periodically through HTTP request. For the same API, Websockets shows the best performance, Sever-Sent-Event is worse, and Long-Polling is the worst. ASRS recommends Websockets by default.
 
--   unit tier (CPU/Memory quota)
+In addition, the cost of message routing also limits the performance. ASRS plays a role of router, which routes the message from a bunch of clients or servers to other clients or servers. Different scenario or API requires different routing. For **echo**, the client sends a message to itself, and the routing destination is also itself. This is the lowest cost routing. But for **broadcast**, **send to group**, **send to connection**, ASRS has to find the target connections through the internal distributed data structure, which consumes more CPU, Memory and even network bandwidth. As a result, performance is not as good as **echo**.
+
+On the default mode, the app server may also become a bottleneck for some scenarios, because Azure SignalR SDK has to invoke the Hub, meanwhile it maintains the connection with every client connection through heart-beat.
+
+On server-less mode, the client sends message by HTTP post which is not as efficient as Websockets.
+
+As a summary, the inbound and outbound capacity is impacted by the following factors:
+
+-   SKU tier (CPU/Memory)
 
 -   connection number
 
 -   message size
 
--   message send/receive rate
+-   message send rate
+
+-   transport type
 
--   use case
+-   use case (routing cost)
 
 -   app server and server connections (on server mode)
 
-Generally, the inbound and outbound message turn less if the sending message size increases.
+<a name="evaluation"></a>
+### Find a proper SKU
+
+How to evaluate the performance inbound/outbound capacity or how to find which tier is suitable for a specific use case?
+
+Let us simplify the evaluation first by assuming some default settings: Websockets is applying, message size is 2048 bytes, sending message every 1 second, and apply default mode.
+
+We assume the app server is powerful enough and is not the performance bottleneck. Then we can check the maximum inbound and outbound bandwidth for every tier.
+
+<a name="quick_evaluation"></a>
+#### Quick evaluation
+
+For the inbound message limit, let us choose **echo** which has the lowest routing cost. For the outbound message limit, broadcast is the good use case. The following table lists the maximum inbound and outbound message bandwidth (bytes per second).
```

Comment:
This is a bit confusing. So the table is actually mixed with echo (for inbound) and broadcast (for outbound)? Better to split into two.

Reply:
Ok.